### DISCUSSION TITLE
proposal: each & every

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#209] [Ideas]
Pattern.every will always apply the function the first time, which is almost never what I want. It would be nice if there was an alternative version that applied the function at the last of n cycles. The most obvious naming would be `each`. e.g.

```js
s("bd").each(4, fast(4))
```

**Comment by yaxu** [comment]
How about finding clearer names, like `firstOf` for every and `lastOf` for each?

Where I've used synonyms in Tidal (e.g. 'bite' and 'chew') I've struggled to remember which is which.

Or maybe there is a more general UI we could find.

Being explicit with `.when("<f f f t>", fast(4))` is clearer and easier to develop/manipulate further, for example.

**Reply by felixroos** [reply]
`firstOf` and `lastOf` sounds good, agreed that it's hard to remember after a while. `every` can stay like it is, so we don't break tidal compatibility.
[.when is already there](https://strudel.tidalcycles.org/tutorial/#patternwhen)

**Reply by yaxu** [reply]
I quite like the idea of changing `every`to `lastof` behaviour. Generally it makes more sense for a variation to happen after a repetition.

I could change this in Tidal 2.0 as well, which will be full of breaking changes.

**Reply by yaxu** [reply]
Yes I mentioned `when` in case a shorter way of describing the same thing is possible. But it's already short enough I guess.

**Reply by yaxu** [reply]
Currently we have two every's:
https://github.com/tidalcycles/strudel/blob/main/packages/core/pattern.mjs#L1010-L1042

I'll remove the first one for now as I think the second one is overriding it.

Shall we go ahead with `firstOf` and `lastOf`? Still unsure about what `every` should do.

**Reply by felixroos** [reply]
> Shall we go ahead with firstOf and lastOf? Still unsure about what every should do.

Yep! I'd say using `lastOf` for `every` makes more sense because it's more common to first establish a pattern and then break it rather than the other way around
